crypto/src/fixpoint.rs

```diff
@@ -36,17 +39,47 @@ impl Debug for U128x128 {
     }
 }
 
-// This is a hack to make the `Display` impl work.
-fn f64_from_secret(value: &U128x128) -> f64 {
+/// Converts an `f64` into a `U128x128`, this method will panic and is lossy.
+/// It should only be used in low stakes client-side contexts.
+/// Chesterton's documentation: `panic!` are to discourage usage of this method.
+pub fn from_f64_unsafe(value: f64) -> U128x128 {
+    if value < 0.0 {
+        panic!("U128x128 cannot represent negative numbers");
+    }
+
+    if value.is_infinite() {
+        panic!("U128x128 cannot represent infinite values");
+    }
+
+    if value.is_nan() {
+        panic!("U128x128 cannot represent NaN values");
+    }
+
+    // Split the f64 into its integral and fractional parts.
+    let integral = value as u128;
+    let fractional = value.fract();
+
+    // Convert the fractional part to an unsigned 128-bit integer.
+    // 1. Multiply the fractional part by 2^128 to move the significant bits to the left.
+    // 2. Round down the result to the nearest integer.
+    // 3. Convert the rounded result to an unsigned 128-bit integer.
+    let fractional_as_u128 = (fractional * f64::from_bits(BASE)).trunc() as u128;
+
+    // Combine the integral and fractional parts into a U128x128.
+    let combined = U256::from_words(integral, fractional_as_u128);
+    U128x128(combined)
+}
+
+/// Converts an `U128x128` into an `f64`, this is a hack to make the `Display`
+/// impl work. This should NOT be used in mission critical contexts.
+pub fn to_f64_unsafe(value: &U128x128) -> f64 {
```

Comment:
There's a conflict since I promoted this to a first-class `From` impl.  I think it's fine to have a conversion; floats aren't exact so shouldn't be used in consensus-critical contexts anyways, but not having conversions at all makes things difficult to use.